src/libraries/System.Text.RegularExpressions/src/dotnet-regular-expressions.md

```diff
@@ -0,0 +1,237 @@
+
+### MatchEvaluator (public)
+
+### RegexCompilationInfo (public)
+
+* Parameters to use for regex compilation to disk
+* Passed in by app to `Regex.CompileToAssembly(..)` - which is not currently implemented.
+
+## Key types - Parsing
+
+### RegexParser
+
+* Converts pattern string to `RegexTree` of `RegexNode`s
+* Invoked with `RegexTree Parse(string pattern, RegexOptions options...) {}`
+* Also has `Escape(..)` and `Unescape(..)` methods, and parses into `RegexReplacement`s
+* Does a partial prescan to prep capture slots
+* As each `RegexNode` is added, it attempts to reduce (optimize) the newly formed subtree. When parsing completes, there is a final optimization of the whole tree.
+
+### RegexReplacement
+
+* Parsed replacement pattern
+* Created by `RegexParser`, used in `Regex.Replace`/`Match.Result(..)`
+
+### RegexCharClass
+
+* Representation of single, range, or class
+* Created by `RegexParser`
+* Creates packed string to be held on `RegexNode`
+* Has utility methods for examining the packed string, in particular for testing membership of the class (`CharInClass(..)`)
+
+### RegexNode
+
+* Node in regex parse tree
+* Created by `RegexParser`
+* Some nodes represent subsequent optimizations, rather than individual elements of the pattern
+* Holds `Children` and `Next`
+* Holds char or string (which may be char class), and `M` and `N` constants (eg loop bounds)
+* Note: polymorphism was not used here: the interpretation of its fields depends on the integer Type field
+
+### RegexTree
+
+* Simple holder for root `RegexNode`, options, and a captures data structure
+* Created by `RegexParser`
+
+### RegexWriter
+
+* Responsible for translating a `RegexTree` to a `RegexCode`
+* Invoked by `Regex`
+* Creates itself `RegexCode Write(RegexTree tree){}`
+
+### RegexFCD
+
+* Responsible for static pattern prefixes
+* Created by `RegexWriter`
+* Creates `RegexFC`s
+* `FirstChars()` creates `RegexPrefix` from `RegexTree`
+* FC means "First chars": not clear what D means...
+
+### RegexPrefix
+
+* Literal string that match must begin with
+
+### RegexBoyerMoore
+
+* Supports searching the text for literals
+* Constructed by `RegexWriter`
+* Singleton held on `RegexCode`
+* `RegexInterpreter` uses it to perform Boyer-Moore search
+* `RegexCompiler` uses the tables from this object, but generates its own code for the Boyer-Moore search
+
+### RegexCode
+
+* Abstract representation of the "program" for a particular pattern
+* Created by `RegexWriter`
+* Code is an array of integers. Within the array, op-codes' types are indicated by integer consts analogous to those on `RegexNode`.
+* Has several related data structures such as a string table, a captures table, and prefixes
+
+## Key types - Compilation (if not interpreted)
+
+### RegexCompiler (public abstract)
+
+* Responsible for compiling `RegexCode` to a `RegexRunnerFactory`
+* As implemented, uses `RegexLWCGCompiler`
```

Comment:
Not understanding this... RegexLWCGCompiler derives from RegexCompiler... RegexCompiler doesn't use RegexLWCGCompiler.

Reply:
Yes, exactly
https://github.com/dotnet/runtime/blob/ed9d30968ffd8f6fb808323941092b9712b9311b/src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexCompiler.cs#L145-L146

In my mind, base classes should not normally have knowledge of their derived classes: they should instead have them passed in or use some kind of factory, or it's too easy to couple everything together otherwise. In this case, since we own the base class, it doesn't really matter.

Reply:
> In my mind, base classes should not normally have knowledge of their derived classes

It's very common for base classes to have factories that can produce derived classes.  XmlReader.Create, for example.